diff --git a/src/crypto/tls/common.go b/src/crypto/tls/common.go
index 5fd92d3c63..8fc665d44a 100644
--- a/src/crypto/tls/common.go
+++ b/src/crypto/tls/common.go
@@ -447,6 +447,9 @@ type ClientHelloInfo struct {
 	// might be rejected if used.
 	SupportedVersions []uint16
 
+	Extensions []uint16
+	Raw        []byte
+
 	// Conn is the underlying net.Conn for the connection. Do not read
 	// from, or write to, this connection; that will cause the TLS
 	// connection to fail.
diff --git a/src/crypto/tls/handshake_messages.go b/src/crypto/tls/handshake_messages.go
index 8620b66a47..87db0dbb45 100644
--- a/src/crypto/tls/handshake_messages.go
+++ b/src/crypto/tls/handshake_messages.go
@@ -77,6 +77,7 @@ type clientHelloMsg struct {
 	compressionMethods               []uint8
 	serverName                       string
 	ocspStapling                     bool
+	extensions                       []uint16
 	supportedCurves                  []CurveID
 	supportedPoints                  []uint8
 	ticketSupported                  bool
@@ -449,6 +450,7 @@ func (m *clientHelloMsg) unmarshal(data []byte) bool {
 		return true
 	}
 
+	m.extensions = []uint16{}
 	var extensions cryptobyte.String
 	if !s.ReadUint16LengthPrefixed(&extensions) || !s.Empty() {
 		return false
@@ -467,6 +469,7 @@ func (m *clientHelloMsg) unmarshal(data []byte) bool {
 			return false
 		}
 		seenExts[extension] = true
+		m.extensions = append(m.extensions, extension)
 
 		switch extension {
 		case extensionServerName:
diff --git a/src/crypto/tls/handshake_server.go b/src/crypto/tls/handshake_server.go
index ac3d915d17..d1ee5dfd30 100644
--- a/src/crypto/tls/handshake_server.go
+++ b/src/crypto/tls/handshake_server.go
@@ -963,6 +963,8 @@ func clientHelloInfo(ctx context.Context, c *Conn, clientHello *clientHelloMsg)
 		SignatureSchemes:  clientHello.supportedSignatureAlgorithms,
 		SupportedProtos:   clientHello.alpnProtocols,
 		SupportedVersions: supportedVersions,
+		Extensions:        clientHello.extensions,
+		Raw:               clientHello.raw,
 		Conn:              c.conn,
 		config:            c.config,
 		ctx:               ctx,
